lib/Sema/CSApply.cpp

```diff
@@ -99,6 +107,69 @@ Solution::computeSubstitutions(GenericSignature sig,
                               lookupConformanceFn);
 }
 
+static clang::FunctionDecl *generateCXXFunctionTemplateSpecializations(
+    ASTContext &ctx, clang::FunctionTemplateDecl *func, SubstitutionMap subst) {
+  ClangTypeConverter typeConv(ctx, func->getASTContext(), llvm::Triple());
+  SmallVector<clang::TemplateArgument, 4> templateSubst;
+  for (clang::NamedDecl *param : *func->getTemplateParameters()) {
+    auto templateParam = cast<clang::TemplateTypeParmDecl>(param);
+    auto replacement = subst.getReplacementTypes()[templateParam->getIndex()];
+    auto qualType = typeConv.convert(replacement);
+    templateSubst.push_back(clang::TemplateArgument(qualType));
+  }
+
+  // Instanciate a specialization of this template using the substitution map.
+  auto *templateArgList = clang::TemplateArgumentList::CreateCopy(
+      func->getASTContext(), templateSubst);
+  auto &sema = ctx.getClangModuleLoader()->getClangInstance().getSema();
+  auto *spec = sema.InstantiateFunctionDeclaration(func, templateArgList,
+                                                   clang::SourceLocation());
+  sema.InstantiateFunctionDefinition(clang::SourceLocation(), spec);
+  return spec;
+}
```

Comment:
What do you think about moving this whole function as a method in ClangModuleLoader?

Reply:
Done.